File: android/app/app/src/main/java/com/loadfinder/app/domain/util/Geo.kt
package com.loadfinder.app.domain.util

import com.loadfinder.app.domain.model.GeoPoint
import kotlin.math.*

object Geo {
    fun distanceKm(a: GeoPoint, b: GeoPoint): Double {
        val r = 6371.0
        val dLat = Math.toRadians(b.lat - a.lat)
        val dLon = Math.toRadians(b.lon - a.lon)
        val x = sin(dLat / 2).pow(2) +
                cos(Math.toRadians(a.lat)) *
                cos(Math.toRadians(b.lat)) *
                sin(dLon / 2).pow(2)
        return 2 * r * atan2(sqrt(x), sqrt(1 - x))
    }
}

File: backend/src/geo.ts
export function distanceKm(a:{lat:number,lon:number}, b:{lat:number,lon:number}) {
  const R=6371, dLat=(b.lat-a.lat)*Math.PI/180, dLon=(b.lon-a.lon)*Math.PI/180;
  const x=Math.sin(dLat/2)**2+
    Math.cos(a.lat*Math.PI/180)*Math.cos(b.lat*Math.PI/180)*Math.sin(dLon/2)**2;
  return 2*R*Math.atan2(Math.sqrt(x),Math.sqrt(1-x));
}